### PR TITLE
Propagate disconnected reason during 'disconnected' event

### PR DESCRIPTION
**Pull request by vivekmogalapalli**
## Description

The container emits a `disconnected` event in event of a connection loss but it doesn't emit any additional details about why the connection was disconnected. This information is already available in the container in the form of `reason` and is also logged in Fluid Framework's telemetry but isn't accessible to the host that integrates with the Fluid Framework. 

This change propagates `reason` when a `disconnected` event is raised by the container. 

## Reviewer Guidance

We initially started off with passing the `reason` as an extra parameter through each of the methods being called as we progressed through the stack 
Here is the [draft PR for the alternative approach](https://github.com/microsoft/FluidFramework/pull/11261). 

The changes in the initial approach above were much more minor but it contrasted with a current design pattern. Currently, the `ConnectionState` is encapsulated within `ConnectionStateHandler` and the `Container` queries the `ConnectionStateHandler` to get the latest connection state in event of a request to propagateConnectionState(). [[Link to the specific location where this happens](https://github.com/microsoft/FluidFramework/blob/ff6850da29c168bfc68c392c5cb934a4855c50bc/packages/loader/container-loader/src/container.ts#L1647)]

Pair programmed with @markfields and introduced `ConnectionStateDetails` to ensure we follow the same design pattern.

## Does this introduce a breaking change?

No

## Other information or known dependencies

- The Microsoft Loop team is relying on this change to measure and improve connectivity reliability across Microsoft Loop experiences
- The **next steps** here would be to instrument in the host experience cases where we don't restore connection within an appropriate time and keep track of the disconnected reason that contribute to the loss of connection

### CODE REVIEW

**Comment by markfields** [comment]
@vladsud @scottn12 - Can you please take a look at this and help Vivek get this in? Thanks!

**Review by scottn12 — APPROVED** [review]
LGTM! 😄

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 1]
```diff
@@ -29,6 +29,16 @@ export interface IConnectionStateHandlerInputs {
...
+/** A type that contains details pertaining to the connection state */
+export type ConnectionStateDetails =
+| {
+    state: ConnectionState.Connected | ConnectionState.CatchingUp;
+}
+| {
+    state: ConnectionState.Disconnected;
+    reason?: string;
+};
+
```
[the remark this answers is not in the document]
Comment (vivekmogalapalli) [reply packages/loader/container-loader/src/connectionStateHandler.ts]: Thanks @scottn12, fixed it!

**Review by WayneFerrao — APPROVED** [review]

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 2]
```diff
@@ -29,6 +29,18 @@ export interface IConnectionStateHandlerInputs {
...
+/** 
+ * A type that contains details pertaining to the connection state 
+ */
+export type ConnectionStateDetails =
+    | {
+        state: ConnectionState.Connected | ConnectionState.CatchingUp;
+    }
+    | {
+        state: ConnectionState.Disconnected;
+        reason?: string;
```
Comment (vladsud) [inline packages/loader/container-loader/src/connectionStateHandler.ts:41]: In general, I find it useful to have that data at different layers.
But, this is an API contract! Meaning that we might be running into various back-compat concerns here (like updating socket.io version may result in different strings, and code breaking in production).
So if you want to add it here, we need a very strong comment here saying that this string can be used only to be logged to telemetry, and that specific values will change over time - none of shipping code can take a dependency on actual values other than logging to telemetry.

**Comment by vladsud** [comment]
I'm not sure I like this approach - I think you are overcomplicating it.
I think it's Ok not to have a reason for disconnected state when container just loads (i.e. this.propagateConnectionState() calls at the end of boot sequence). If that's the case, then propagating disconnect reason for all disconnects becomes trivial and does not need this change. All you need is something I'm doing here (this is on top of Mark's PR that is already merged, i.e. only last commit is interesting) -https://github.com/microsoft/FluidFramework/pull/10958
Basically merging connectionStateChanged and logConnectionStateChangeTelemetry callbacks together, and you get a reason to propagate!

If I got it right (and it addresses you concerns), I'd rather go this way, as it's much simpler, does not add new concepts / tracking.

**Comment by markfields** [comment]
@vladsud - Let me chime in and respond to your overall feedback.

> I'm not sure I like this approach - I think you are overcomplicating it.

If this is overcomplicated, I'm to blame :)

> Basically merging connectionStateChanged and logConnectionStateChangeTelemetry callbacks together, and you get a reason to propagate!

I agree we could merge those two, but let's set that aside since it's extra beyond what Vivek needs.

I think what you're saying is that we can easily just pass `reason` up through `Container.propagateConnectionState` via `connectionStateChanged` callback.  This simple solution is precisely what Vivek started with - see #11261.

When he reviewed that privately with me, here was my concern that lead me to recommend this more complex solution -- If you look at `Container.propagateConnectionState` in that other PR, you'll notice that `state` and `reason` are decoupled.  `state` is read off `ConnectionStateHandler`, but `reason` is passed in.  In practice, it's fine, they are correlated.  But I thought the approach in this PR was more rigorous/correct - it keeps `state` and `reason` coupled together in one object.

If we want to set aside this design concern, then I'm fine with the simpler PR.  In fact I'm signed off on it now.  @vivekmogalapalli if you don't mind waiting for Vlad to circle back and respond to my comments here, then you can decide which route to take based on his feedback.

Thanks for jumping in and making this change!  Definitely good to plumb this info through!

File: packages/loader/container-loader/src/connectionStateHandler.ts (hunk 2 above)
Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts:41]: +1

Please do include some documentation explaining that this is for diagnostic/telemetry purposes only. Any exported API with this parameter should have doc comments explaining this, and the README should be updated to note that the disconnected event may have a string argument which is this reason, to be used only for diagnostic purposes.

**Comment by vladsud** [comment]
@markfields, @vivekmogalapalli, if you really want to go with keeping disconnected state around, then I'd suggest decoupling it out of ConnectionState into separate variable - same as we track clientId.
Notice that we have ConnectionState.EstablishingConnection (unused, but I assume we want to use it). There might be more states added in the future. It will be pain in the butt to maintain such complex data structures / transitions. It's much easier to say - we always keep disconnect reason, and reset it on disconnect event, and communicate on these 2-3 events.

While being very loosely related, worth pointing out that I raised https://github.com/microsoft/FluidFramework/pull/11895 that merges ConnectionStateHandler callbacks and communicates disconnect reason. If I were you, I'd go with very simple solution - not keeping any extra state, adding reason to disconnect event, and not care about having right reason at the end of boot sequence (this only matters if we were able to connect and disconnect while loading, which is not applicable to Bohemia that delays connectiosn).

**Comment by vivekmogalapalli** [comment]
Thanks for the feedback everyone. I'm closing out this PR in favor of the simpler PR since this change is primarily for telemetry purposes and to keep changes to a minimum.
